File: 198/Soluzione.cpp
#include<vector>
#include<iostream>
using namespace std;
#define N 200

class Solution {
public:
    int rob(vector<int>& nums) {
        return max(Recursive(nums, 0),Recursive(nums, 1));
    }
    int Recursive(vector<int> nums, int index) {
        if(index > nums.size()-1)
            return 0;
        return max(nums[index] + Recursive(nums, index+2), nums[index] + Recursive(nums, index+3));
    }
};

int main() {
    Solution soluzione;
    vector<int> nums;
    for(int i = 0; i < N; i++)
        nums.push_back(i);

    cout << soluzione.rob(nums);

}

File: LeetCode/DP/198/Soluzione.java
import java.util.Arrays;

class Solution {
    public int rob(int[] nums) {
        int cache[] = new int[nums.length];
        Arrays.fill(cache, -1);
        return DP(nums, 0, cache);
    }

    public int DP(int[] nums, int index, int[] cache) {
        if(index >= nums.length)
            return 0;

        if(cache[index] >= 0)
            return cache[index];

        int result = Math.max((nums[index] + DP(nums, index + 2, cache)), DP(nums, index + 1, cache));
        cache[index] = result;
        return result;
    }

    public static void main(String[] arg) {
        int nums[] = {1, 2, 3, 1};
        Solution soluzione = new Solution();
        System.out.println(soluzione.rob(nums));
    }
}